test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs

```diff
@@ -1791,7 +1791,7 @@ public async Task RestoreNetCore_MultipleProjects_SameToolDifferentVersionsAsync
                 var zPath = Path.Combine(pathContext.UserPackagesFolder, ".tools", "z");
 
                 // Act
-                var r = Util.RestoreSolution(pathContext);
+                var r = Util.RestoreSolution(pathContext, expectedExitCode: 0, "-DisableParallelProcessing");
```

Comment:
I'm not sure how this is supposed to help. 

The failure suggests the package is *not* installed in the expected location. 
Can we maybe try to dump the directory structure of the GPF so we can better understand what's happening?

Reply:
Error output shown below (edited for better readability). I suspect some process/thread is looking for the cache file before being created.

```
Could not find file 'C:\A\1\1\s\.test\work\5ef9336e\878364fa\globalPackages\.tools\z\20.0.0\netcoreapp1.0\z.nuget.cache'.
System.IO.FileNotFoundException: Could not find file 'C:\A\1\1\s\.test\work\5ef9336e\878364fa\globalPackages\.tools\z\20.0.0\netcoreapp1.0\z.nuget.cache'.
File name: 'C:\A\1\1\s\.test\work\5ef9336e\878364fa\globalPackages\.tools\z\20.0.0\netcoreapp1.0\z.nuget.cache'
   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.FileStream.Init(String path, FileMode mode, FileAccess access, Int32 rights, Boolean useRights, FileShare share, Int32 bufferSize, FileOptions options, SECURITY_ATTRIBUTES secAttrs, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share)
   at NuGet.Common.FileUtility.SafeRead[T](String filePath, Func`3 read)
   at NuGet.Commands.RestoreCommand.EvaluateCacheFile()
   at NuGet.Commands.RestoreCommand.<ExecuteAsync>d__33.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.Commands.RestoreRunner.<ExecuteAsync>d__7.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.Commands.RestoreRunner.<ExecuteAndCommitAsync>d__6.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.Commands.RestoreRunner.<CompleteTaskAsync>d__9.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at NuGet.Commands.RestoreRunner.<RunAsync>d__2.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.Commands.RestoreRunner.<RunAsync>d__0.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.Commands.RestoreRunner.<RunAsync>d__1.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.CommandLine.RestoreCommand.<ExecuteCommandAsync>d__52.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.CommandLine.Command.Execute()
   at NuGet.CommandLine.Program.MainCore(String workingDirectory, String[] args)


   NuGet Version: 5.9.0.13210
Restoring NuGet packages for solution C:\A\1\1\s\.test\work\5ef9336e\878364fa\solution\solution.sln.
MSBuild auto-detection: using msbuild version '16.7.0.37604' from 'C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\bin'. Use option -MSBuildVersion to f
```

Reply:
Ah, than maybe the safe reading utility is not that safe!

I think I understand the problem. 

We might need to tweak with the permissions we use when we write/read the cache file. 

The LockFile suffers from this as well. 

https://github.com/NuGet/NuGet.Client/blob/f88ad395c062c98a60ba3084e99f5476106e00d0/src/NuGet.Core/NuGet.ProjectModel/LockFile/LockFileUtilities.cs#L23

Reply:
I am wondering if this is because of our recent change to use `MemoryMappedFile` in [PR](https://github.com/NuGet/NuGet.Client/pull/3524). The first comment in that PR says 

`You'll notice a bunch of "file already in use" errors in the test failures in CI. These seem to be spurious/random and I'm not sure exactly what's triggering them but I can probably work around it with a try{} around the mmap code that falls through to a regular write. These are rare enough that it shouldn't affect performance, and if we catch the right exception, won't affect correctness (because the incoming stream won't have been consumed at all).`

One of my private builds reported the same error message. I checked the folder structure and found that the file exists in location. `C:\repos\NuGet.Client\.test\work\5ef9336e\878364fa\globalPackages\.tools\z\20.0.0\netcoreapp1.0\z.nuget.cache`. The directory also had `assets.json` file.

Reply:
I don't know the entire restore workflow, hence tried to reproduce this issue locally. After couple of attempts I can reproduce it. 
![image](https://user-images.githubusercontent.com/52756182/95020255-ebd01400-061e-11eb-98da-b13c80b0b535.png)
There is a race condition [here](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Common/PathUtil/FileUtility.cs#L213). If a thread deletes a file, then other reads which are trying to [read](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Common/PathUtil/FileUtility.cs#L233) the file throw file not found exception. 

Here is the call stack for the thread that deleted the file before other threads read the same file.

```cs
NuGet.Common.dll!NuGet.Common.FileUtility.Delete(string path) Line 156	C#
NuGet.Common.dll!NuGet.Common.FileUtility.Replace(string sourceFileName, string destFileName) Line 159	C#
NuGet.Common.dll!NuGet.Common.FileUtility.Replace(System.Action<string> writeSourceFile, string destFilePath) Line 104	C#	NuGet.Commands.dll!NuGet.Commands.RestoreResult.CommitAssetsFileAsync(NuGet.ProjectModel.LockFileFormat lockFileFormat, NuGet.Commands.IRestoreResult result, NuGet.Common.ILogger log, bool toolCommit, System.Threading.CancellationToken token) Line 230	C#
NuGet.Commands.dll!NuGet.Commands.RestoreResult.CommitAsync(NuGet.Common.ILogger log, System.Threading.CancellationToken token) Line 165	C#
NuGet.Commands.dll!NuGet.Commands.RestoreRunner.CommitAsync(NuGet.Commands.RestoreResultPair restoreResult, System.Threading.CancellationToken token) Line 263	C# 	NuGet.Commands.dll!NuGet.Commands.RestoreRunner.ExecuteAndCommitAsync(NuGet.Commands.RestoreSummaryRequest summaryRequest, System.Threading.CancellationToken token) Line 233	C#
NuGet.Commands.dll!NuGet.Commands.RestoreRunner.RunAsync.AnonymousMethod__0() Line 84	C#
```
I think the easiest fix would be to remove [`FileShare.Delete`](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Common/PathUtil/FileUtility.cs#L17) so that `Delete` operation will make [three](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Common/PathUtil/FileUtility.cs#L16) attempts to delete the file allowing other threads to safely read the file and then throw an exception. 
If other team members think that this is a bug in the restore logic, I am happy to engage in further conversation and learn more.

Reply:
The bug is definitely in our read/write utility. It's manifesting in restore because we have *multiple* threads trying to access the same file, but the root cause is isolated to the utility.

I think your suggestion might make sense, but given that it's the file utility that we're changing we need to consider the overall impact. 

Like we'd need to go back through the issues/PRs to understand why the `ReadWrite | Delete` perms were used. 

We don't want to regress other scenarios here :) 

Given the uncommon scenario (the scenario relies on a missing version for DotnetCliTool), and the lack of customer reports, I'd lean towards disabling/delete/disabling parallel processing in this test, over implementing an actual fix. 

I think I'd prefer an issue just sitting in the icebox and wait for us to hear an actual customer complaint. 

Happy to discuss though.